lmdeploy/vl/engine.py

```diff
@@ -135,36 +140,53 @@ async def _forward_loop(self):
                 while self._que.qsize() == 0:
                     await asyncio.sleep(0)
                 item = await self._que.get()
-                record.enqueue(item[0], item[1])
-            inputs = record.dequeue(self.max_batch_size)
+                record.enqueue(item[0], item[1], item[2])
+            inputs, kwargs = record.dequeue(self.max_batch_size)
             future = asyncio.get_event_loop().run_in_executor(
-                None, self.forward, inputs)
+                None, self.forward, inputs, kwargs)
             future.add_done_callback(_raise_exception_on_finish)
             outputs = await future
             record.done.extend(outputs)
             while record.notify():
                 pass
 
-    def forward(self, inputs: List[Image]):
+    def _init_input_kwargs(self,
+                           inputs: List[Image],
+                           kwargs: List[Dict] = None):
+        """Check and init inputs kwargs."""
+        if kwargs is None:
+            kwargs = [{}] * len(inputs)
+        assert len(kwargs) == len(inputs), \
+            'different length of inputs and kwargs'
+        return kwargs
+
+    def forward(self, inputs: List[Image], kwargs: List[Dict] = None):
         """Model forward."""
+        kwargs = self._init_input_kwargs(inputs, kwargs)
         time_start = time.perf_counter()
-        outputs = self.model.forward(inputs)
+        func_params = inspect.signature(self.model.forward).parameters
+        func_inputs = [inputs, kwargs] if len(func_params) > 1 else [inputs]
+        outputs = self.model.forward(*func_inputs)
         if isinstance(outputs[0], torch.Tensor):
             outputs = [x.cpu() for x in outputs]
         time_end = time.perf_counter()
         logger.info(f'ImageEncoder forward {len(inputs)} images, '
                     f'cost {time_end - time_start:.3f}s')
         return outputs
 
-    def infer(self, inputs: List[Image]):
+    def infer(self, inputs: List[Image], kwargs: List[Dict] = None):
```

Comment:
I don't appreciate the `kwargs` and `_init_input_kwargs`.
They make the code hard to read